### PR TITLE
Add codeModulesImage field to Dynakube where CSI Driver is used

### PR DESCRIPTION
**Pull request by waodim**
# Description

Added a image field in the dynakube with the modes enabled, which use the CSIDriver. The field is called "codeModulesImage".

## How can this be tested?
Apply a dynakube with either cloudNativeFullStack enabled with the field "codeModulesImage" set or with applicationMonitoring enabled with the field "codeModulesImage" set. If one decides to apply a dynakube with applicationMonitoring enabled, useCSIDriver field set to false and with the "codeModulesImages" field set, the validation webhook responds with an error.

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly

### CODE REVIEW

**Comment by luhi-DT** [comment]
I think you forgot to add general files like CRD and so on

File: src/api/v1beta1/oneagent_types.go [hunk 1]
```diff
@@ -29,6 +29,10 @@ type OneAgentSpec struct {
 }
 
 type CloudNativeFullStackSpec struct {
+	// Optional: the Dynatrace installer container image
+	// Defaults to the registry on the tenant for both Kubernetes and for OpenShift
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Image",order=12,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:advanced","urn:alm:descriptor:com.tectonic.ui:text"}
```
Comment (0sewa0) [inline src/api/v1beta1/oneagent_types.go:34]:
```suggestion
	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="CodeModulesImage",order=12,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:advanced","urn:alm:descriptor:com.tectonic.ui:text"}
```

File: src/api/v1beta1/oneagent_types.go [hunk 2]
```diff
@@ -115,6 +119,11 @@ type HostInjectSpec struct {
 type ApplicationMonitoringSpec struct {
 	AppInjectionSpec `json:",inline"`
 
+	// Optional: the Dynatrace installer container image
+	// Defaults to the registry on the tenant for both Kubernetes and for OpenShift
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Image",order=12,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:advanced","urn:alm:descriptor:com.tectonic.ui:text"}
```
Comment (0sewa0) [inline src/api/v1beta1/oneagent_types.go:124]:
```suggestion
	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="CodeModulesImage",order=12,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:advanced","urn:alm:descriptor:com.tectonic.ui:text"}
```

File: src/api/v1beta1/properties.go [hunk 3]
```diff
@@ -166,6 +166,10 @@ func (dk *DynaKube) Image() string {
 		return dk.Spec.OneAgent.ClassicFullStack.Image
 	} else if dk.HostMonitoringMode() {
 		return dk.Spec.OneAgent.HostMonitoring.Image
+	} else if dk.CloudNativeFullstackMode() {
+		return dk.Spec.OneAgent.CloudNativeFullStack.CodeModulesImage
```
Comment (0sewa0) [inline src/api/v1beta1/properties.go:170]: shouldn't these be in there own func ?  `func (dk *DynaKube) CodeModulesImage() string {...}`
there will be an image field for cloudNative that will use this func

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/api/v1beta1/properties.go (hunk 3 above)
Comment (waodim) [reply src/api/v1beta1/properties.go:170]: you are right, did not think about that. thank you.

File: src/api/v1beta1/properties.go [hunk 4]
```diff
@@ -225,6 +234,11 @@ func (dk *DynaKube) ImmutableOneAgentImage() string {
 		return oneAgentImage // TODO: What to do with the Version field in this case ?
 	}
 
+	codeModulesImage := dk.CodeModulesImage()
```
Comment (0sewa0) [inline src/api/v1beta1/properties.go:237]: do we need this ? @chrismuellner

Comment (luhi-DT) [reply src/api/v1beta1/properties.go:237]: I dont think we need that here, CodeModuleImage should be an own function and has nothing in common with the 'image' field of the osAgent

Comment (waodim) [reply src/api/v1beta1/properties.go:237]: okay, yes that makes sense. what i do not understand currently is, where is CodeModulesImage() then called? Or is it nowhere called for the moment?

Comment (chrismuellner) [reply src/api/v1beta1/properties.go]: that's fine, it will be used by the csi driver in the future

File: src/api/v1beta1/oneagent_types.go [hunk 5]
```diff
@@ -29,6 +29,11 @@ type OneAgentSpec struct {
 }
 
 type CloudNativeFullStackSpec struct {
+	// Optional: the Dynatrace installer container image
```
Comment (chrismuellner) [inline src/api/v1beta1/oneagent_types.go:32]: the description isn't quite accurate:
- this doesn't default to anything, but needs to be set by the user to a private registry

**Review by chrismuellner — CHANGES_REQUESTED** [review]

File: src/api/v1beta1/oneagent_types.go [hunk 6]
```diff
@@ -29,6 +29,11 @@ type OneAgentSpec struct {
...
+	// Defaults to the registry on the tenant for both Kubernetes and for OpenShift
+	// +operator-sdk:csv:customresourcedefinitions:type=spec,displayName="CodeModulesImage",order=12,xDescriptors={"urn:alm:descriptor:com.tectonic.ui:advanced","urn:alm:descriptor:com.tectonic.ui:text"}
+	CodeModulesImage string `json:"codeModulesImage,omitempty"`
```
Comment (0sewa0) [inline src/api/v1beta1/oneagent_types.go:35]: couldn't this be in the `AppInjectionSpec` ?

File: src/api/v1beta1/oneagent_types.go (hunk 5 above)
Comment (waodim) [reply src/api/v1beta1/oneagent_types.go:32]: ah yes thank you for reminding me.

File: src/api/v1beta1/oneagent_types.go (hunk 6 above)
Comment (waodim) [reply src/api/v1beta1/oneagent_types.go:35]: so you basically mean to remove the field out of the ApplicationMonitoringSpec and the CloudNativeFullStackSpec and just add it to the AppInjectionSpec?

Comment (0sewa0) [reply src/api/v1beta1/oneagent_types.go:35]: yas :)

**Comment by 0sewa0** [comment]
run `make manifests` to update the the yamls according to the recent changes 😉

**Review by 0sewa0 — APPROVED** [review]

**Pull request merged**